app/code/Magento/Catalog/Model/Product/SalabilityChecker.php

```diff
@@ -54,4 +63,23 @@ public function isSalable($productId, $storeId = null): bool
 
         return $product->isSalable();
     }
+    
+    /**
+     * Check if product is visible on frontend.
+     *
+     * @param int|string $productId
+     * @param int|null $storeId
+     * @return bool
+     */
+    public function isProductVisible($productId, $storeId = null): bool
```

Comment:
Unfortunately adding new methods to existing class - backward incompatible change. please move this logic inside constructor

Reply:
I have updated code according to backward compatibility.